### PR TITLE
lecture 3 tasks

### PR DESCRIPTION
**Pull request by litvinyuk852**

### CODE REVIEW

File: Lecture3/Task2/GroupStudent.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            if (filterLastName.startsWith(filterLetter)) {
+                System.out.println(list.get(i));
+            }
+        }
+        System.out.println("");
+    }
+
+    public static void averageAge(List<Student> list) {
+        int sum = 0;
+        for (int i = 0; i < list.size(); i++) {
```
Comment (acherkas) [inline Lecture3/Task2/GroupStudent.java]: for (Student aStudentList : studentList) {
sumAge += aStudentList.getAge();
}

File: Lecture3/Task2/GroupStudent.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Map<Integer, String> groupStudentsMap = new HashMap<>();
+
+        for (int i = 0; i < list.size(); i++) {
+            groupStudentsMap.put(list.get(i).getId(), list.get(i).getLastName());
+        }
+        System.out.println("Group students map: " + groupStudentsMap);
+        return groupStudentsMap;
+    }
+
+    private static void filterMapById(List<Student> list, int filterNum) {
```
Comment (acherkas) [inline Lecture3/Task2/GroupStudent.java]: naming extraction

File: Lecture3/Task3/GroupStudentStream.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class GroupStudentStream {
+    public static void main(String[] args) {
+        List<Student> groupStudents = new ArrayList<>();
```
Comment (acherkas) [inline Lecture3/Task3/GroupStudentStream.java]: the same Builder or Factory pattern

File: Lecture3/Task3/GroupStudentStream.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        sortByAgeStream(groupStudents);
+        filterLastNameStream(groupStudents, "S");
+        averageAgeStream(groupStudents);
+        filterMapByIdStream(groupStudents, 100);
+
+    }
+
+
+    public static void sortByAgeStream(List<Student> list) {
```
Comment (acherkas) [inline Lecture3/Task3/GroupStudentStream.java]: the same as previous

File: Lecture3/Task4/TimeEstimation.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        executionTimeAddToArrayList(arrayList, 1000);
+        executionTimeAddToLinkedList(linkedList, 1000);
+        executionTimeRemoveFromArrayList(arrayList, 1000);
+        executionTimeRemoveFromLinkedList(linkedList, 1000);
+        executionTimeSearchInArrayList(arrayList, 1000);
+        executionTimeSearchInLinkedList(linkedList, 1000);
+
+    }
+    private static void executionTimeAddToArrayList (ArrayList<Integer> arrayList, int elementNum) {
```
Comment (acherkas) [inline Lecture3/Task4/TimeEstimation.java]: extract to util or helpers class

File: Lecture3/Task1/DiskCreation.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        filtered = filtered.stream()
+                .filter(rockSong -> rockSong.getArtist().startsWith(artist))
+                .filter(rockSong -> rockSong.getListeningAmount() > amount)
+                .collect(Collectors.toList());
+        filtered.forEach(System.out::println);
+        System.out.println("");
+    }
+
+    @Override
+    public void printDiskCover() {
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: fixed

File: Lecture3/Task1/DiskCreation.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DiskCreation implements DiskCover {
+    public static void main(String[] args) {
+        List<RockSong> rockSongs = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: factory pattern was applied

File: Lecture3/Task1/DiskCreation.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        Collections.addAll(rockSongs, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25);
+        printTrackList(rockSongs, 20);
+        totalMin(rockSongs, 20);
+        trackListSortByTime(rockSongs, 20);
+        filterTrackByArtistAndListeningAmount(rockSongs, 20, "H", 200);
+
+
+    }
+
+    public static List getTrackList(List<RockSong> list, int numTracks) {
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: method naming was changed

[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: methods are extracted to CommonUtilsDiskCreation class

File: Lecture3/Task1/DiskCreation.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        Comparator<RockSong> byListeningAmount = Comparator.comparingInt(RockSong::getListeningAmount);
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: fixed

File: Lecture3/Task1/DiskCreation.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        list = list.stream()
+                .sorted(byListeningAmount)
+                .collect(Collectors.toList());
+        List<RockSong> trackList = list.subList(0, numTracks);
+        return trackList;
+    }
+
+    public static void printTrackList(List<RockSong> list, int numTracks) {
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: methods are extracted to CommonUtilsDiskCreation class

File: Lecture3/Task1/DiskCreation.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    public static void totalMin(List<RockSong> list, int numTracks) {
+        List<RockSong> trackList = getTrackList(list, numTracks);
+        Integer totalMin = trackList.stream()
+                .mapToInt(RockSong::getTime)
+                .sum();
+        System.out.println("Total min: " + totalMin);
+        System.out.println("");
+    }
+
+    public static void trackListSortByTime(List<RockSong> list, int numTracks) {
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: method naming was changed

File: Lecture3/Task1/DiskCreation.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: naming was changed

File: Lecture3/Task1/DiskCreation.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        Comparator<RockSong> byTime = Comparator.comparingInt(RockSong::getTime);
+        List<RockSong> sortedByTimeList = getTrackList(list, numTracks);
+        sortedByTimeList = sortedByTimeList.stream()
+                .sorted(byTime)
+                .collect(Collectors.toList());
+        sortedByTimeList.forEach(System.out::println);
+        System.out.println("");
+    }
+
+    public static void filterTrackByArtistAndListeningAmount(List<RockSong> list, int numTracks, String artist, int amount) {
```
[the remark this answers is not in the document]
Comment (litvinyuk852) [reply Lecture3/Task1/DiskCreation.java]: method naming was changed

**Pull request merged**